torch/csrc/jit/passes/concat_opt.cpp

```diff
@@ -0,0 +1,487 @@
+#include <torch/csrc/jit/passes/concat_opt.h>
+
+#include <algorithm>
+#include <unordered_set>
+#include <vector>
+
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/named_value.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+void removeCatNodeFromGraph(Node* n) {
+  TORCH_INTERNAL_ASSERT(n->kind() == aten::cat);
+  auto inp_list = n->input(0);
+  GRAPH_UPDATE("Deleting\n", *n);
+  n->destroy();
+  if (!inp_list->hasUses()) {
+    GRAPH_UPDATE("Deleting\n", *inp_list->node());
+    inp_list->node()->destroy();
+  }
+}
+
+bool equal(at::ArrayRef<Value*> list1, at::ArrayRef<Value*> list2) {
+  return list1.size() == list2.size() &&
+      std::equal(list1.begin(), list1.end(), list2.begin());
+}
+
+class ConcatCommonInputsEliminator {
+ public:
+  explicit ConcatCommonInputsEliminator(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  void run() {
+    handleBlock(graph_->block());
+    postprocess();
+  }
+
+ private:
+  void handleBlock(Block* block) {
+    for (auto node : block->nodes()) {
+      if (node->kind() == aten::cat) {
+        handleCat(node);
+      }
+      for (Block* block : node->blocks()) {
+        handleBlock(block);
+      }
+    }
+  }
+
+  void handleCat(Node* node) {
+    GRAPH_DEBUG("Considering cat node for CSE opt: ", node);
+
+    // Do not optimize cat nodes whose inputs are mutated in the graph.
+    // TODO: Improve this by checking if it is mutated in the graph region
+    // where this optimization is applied.
+    if (getOrCreateAliasDb()->hasWriters(node->input(0))) {
```

Comment:
Good point. I have added a check to not consider any cat node whose output has writes for future optimizations. Note that that particular cat (which has a write) could itself be optimized using some previous cat. That is okay since the output of this is never used for this optimizations.